### PR TITLE
New Importer: evalita2009, SIWIS,MSPKA + utilities

### PR DESCRIPTION
**Pull request by eziolotta**
Hi,
this is  implementation of  **Evalita2009, SIWIS Database, MSPKA Corpus**  importer - see [#90](https://github.com/MozillaItalia/DeepSpeech-Italian-Model/issues/90)  

1) Importers extend the ArchiveImporter utility in **corpora_importer.py**

ArchiveImporter implements download + extract + resample mp3 / wav-> wav + csv (wav and csv is compatible with DeepSpeech training)

Importers contains only parsing logic of transcripts and searching for the corresponding audio path.
I tested on Windows 10, Python 3.6.5.

2) Utility **corpora_collector.py** takes care of collecting imported data, by building a new aggregate dataset.

one test with corpora_collector and these 3 importers produce a total of 12.18h of audio divided into 9882 clips lasting less than 15 seconds.

### CODE REVIEW

File: MITADS-Speech/corpora_importer.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        ## audios is a list of all audio_file , full absolute path
+        ## datasets_sizes: dimension of train-test-dev datasets (in different csv)
+        self.utterences = utterences
+        self.audios = audios
+        self.datasets_sizes = datasets_sizes
+        self.make_wav_resample = make_wav_resample
+
+
+class ArchiveImporter:
+    def __init__(self,corpus_name,archive_url,data_dir=None,csv_append_mode=False):
```
Comment (nefastosaturo) [inline MITADS-Speech/corpora_importer.py:76]:
```suggestion
    def __init__(self,corpus_name,archive_url,extract_dir=None, data_dir=None,csv_append_mode=False):
```

File: MITADS-Speech/corpora_importer.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        self.corpus_name=corpus_name
+        self.archive_url=archive_url
+        ##Make archive_name from archive_filename
+        archive_filename = self.archive_url.rsplit('/', 1)[-1]    
+        self.archive_name = archive_filename.rsplit('.', 1)[0]
```
Comment (nefastosaturo) [inline MITADS-Speech/corpora_importer.py:81]:
```suggestion
        self.archive_name = archive_filename.rsplit('.', 1)[0]
        self.extract_dir = self.archive_name
        if extract_dir is not None:
            self.extract_dir = extract_dir        
```

File: MITADS-Speech/corpora_importer.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+        if not path.exists(self.dataset_output_path):
+            print('No path "%s" - creating ...' % self.dataset_output_path)
+            makedirs(self.dataset_output_path)
+
+        archive_filename = self.archive_url.rsplit('/', 1)[-1]
+        # Conditionally download data
+        archive_path = maybe_download(archive_filename, self.origin_data_path, self.archive_url)
+        # Conditionally extract common voice data
+        self._maybe_extract(self.origin_data_path, self.archive_name, archive_path)
```
Comment (nefastosaturo) [inline MITADS-Speech/corpora_importer.py:103]:
```suggestion
        self._maybe_extract(self.origin_data_path,self.extract_dir, archive_path)
```

File: MITADS-Speech/corpora_importer.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        ##get corpus:  audio_file_names + transcriptions
+        print('Filter audio file and parse transcript...')
+        corpus = self.get_corpus()
+
+         # Conditionally convert CSV files and mp3/wav data to DeepSpeech CSVs and wav
+        self._maybe_convert_sets(corpus)      
+
+    def _maybe_extract(self,target_dir, extracted_data, archive_path):
+        # If target_dir/extracted_data does not exist, extract archive in target_dir
+        extracted_path = os.path.join(target_dir, extracted_data)
```
Comment (nefastosaturo) [inline MITADS-Speech/corpora_importer.py:114]:
```suggestion
```suggestion
    def _maybe_extract(self,target_dir, extract_dir, archive_path):
        # If target_dir/extract_dir does not exist, extract archive in target_dir
        extracted_path = os.path.join(target_dir, extract_dir)
```
```

File: MITADS-Speech/siwis_importer.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+class SiwisImporter(ArchiveImporter):
+
+
+    def get_corpus(self):
+        ##extract training and development datasets
+        ##do data merge, ArchiveImporter make final train/test/dev datasets
+        utterances = {}
+        audios = []
+        wav_dir  = os.path.join(self.origin_data_path, self.archive_name, "wav","IT")
+        text_dir = os.path.join(self.origin_data_path, self.archive_name, "text","IT")
```
Comment (nefastosaturo) [inline MITADS-Speech/siwis_importer.py:19]:
```suggestion
        wav_dir  = os.path.join(self.origin_data_path, self.extract_dir, "wav","IT")
        text_dir = os.path.join(self.origin_data_path, self.extract_dir, "text","IT")
```

File: MITADS-Speech/siwis_importer.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        corpus.make_wav_resample = True
+        return corpus
+
+
+if __name__ == "__main__":
+
+    corpus_name=CORPUS_NAME
+    archivie_url = 'https://phonogenres.unige.ch/downloads/siwis_latest.zip'
+    data_dir=None
+    siwis_importer = SiwisImporter(corpus_name,archivie_url,  data_dir=data_dir)
```
Comment (nefastosaturo) [inline MITADS-Speech/siwis_importer.py:63]:
```suggestion
    corpus_name=CORPUS_NAME
    archive_url = 'https://phonogenres.unige.ch/downloads/siwis_latest.zip'
    data_dir=None
    siwis_importer = SiwisImporter(corpus_name,archive_url,extract_dir="siwis_database", data_dir=data_dir)
```

**Review by nefastosaturo — CHANGES_REQUESTED** [review]
from siwis_latest.zip the extracted folder name is siwis_database. the _maybe functions will look to siwis_latest folder that will be always missed

**Comment by Mte90** [comment]
@nefastosaturo can you check this again?

File: MITADS-Speech/corpora_importer.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        self.make_wav_resample = make_wav_resample
+
+
+class ArchiveImporter:
+    def __init__(self,corpus_name,archive_url,extract_dir=None, data_dir=None,csv_append_mode=False):
+        self.corpus_name=corpus_name
+        self.archive_url=archive_url
+        ##Make archive_name from archive_filename
+        archive_filename = self.archive_url.rsplit('/', 1)[-1]    
+        self.archive_name = archive_filename.split('.', 1)[0]## FIX: archive name with extension .tar.gz
```
Comment (nefastosaturo) [inline MITADS-Speech/corpora_importer.py:81]:
```suggestion
        # Make archive_name from archive_filename
        archive_filename = self.archive_url.rsplit('/', 1)[-1]
        # os.path.splitext:
        # tar.gz: will split ("file.name.tar",".gz")
        # but will split correctly "cnz_1.0.0.zip" into ("cnz_1.0.0","zip")
        self.archive_name = os.path.splitext(archive_filename)[0]
        if self.archive_filename.endswith(".tar"):
          self.archive_name = self.archive_name.replace(".tar","")
```

File: MITADS-Speech/siwis_importer.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+class SiwisImporter(ArchiveImporter):
+
+
+    def get_corpus(self):
+        ##extract training and development datasets
+        ##do data merge, ArchiveImporter make final train/test/dev datasets
+        utterances = {}
+        audios = []
+        wav_dir  = os.path.join(self.origin_data_path, self.extract_dir, "wav","IT")
+        text_dir = os.path.join(self.origin_data_path, self.extract_dir, "text","IT")
```
Comment (nefastosaturo) [inline MITADS-Speech/siwis_importer.py:20]:
```suggestion
        text_dir = os.path.join(self.origin_data_path, self.extract_dir, "txt","IT")
```

File: MITADS-Speech/requirements.txt [hunk 9]
```diff
@@ -1,2 +1,6 @@
 BeautifulSoup4
-requests
\ No newline at end of file
+requests
+sox
```
Comment (nefastosaturo) [inline MITADS-Speech/requirements.txt:3]:
```suggestion
sox
charset_normalizer
```

**Review by nefastosaturo — COMMENTED** [review]
add also:

sudo apt install sox

**Pull request merged**